compiler/rustc_hir/src/lang_items.rs

```diff
@@ -55,21 +52,21 @@ macro_rules! language_item_table {
     (
         $( $(#[$attr:meta])* $variant:ident, $module:ident :: $name:ident, $method:ident, $target:expr, $generics:expr; )*
     ) => {
-
-        rustc_data_structures::enum_from_u32! {
-            /// A representation of all the valid lang items in Rust.
-            #[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, Encodable, Decodable)]
-            pub enum LangItem {
-                $(
-                    #[doc = concat!("The `", stringify!($name), "` lang item.")]
-                    ///
-                    $(#[$attr])*
-                    $variant,
-                )*
-            }
+        /// A representation of all the valid lang items in Rust.
+        #[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, Encodable, Decodable)]
+        pub enum LangItem {
+            $(
+                #[doc = concat!("The `", stringify!($name), "` lang item.")]
+                $(#[$attr])*
+                $variant,
+            )*
         }
 
         impl LangItem {
+            const fn items() -> [LangItem; std::mem::variant_count::<LangItem>()] {
+                [$(LangItem::$variant,)*]
```

Comment:
Can we add an assertion somewhere that `LangItem::items()[lang_item as usize] == lang_item`. Just in case something changes with discriminant assignment.

Reply:
I went to do that, and then I wrote a comment explaining the assumption, and then I decided to revert to the original branch-based function, because eliminating the macro is the main point of this change. Someone can do a follow-up if they like to do the array thing. I did add a comment explaining that the branch-based function doesn't make any assumptions about discriminant ordering.

Reply:
That works for me.